### DISCUSSION TITLE
Create an API for my app

### DISCUSSION THREAD

**Discussion by Ethanol48** [pocketbase/pocketbase#5363] [Q&A] [answered]
I'm building an app and I would like to create a custom API for my customers, now I have a problem, I don't know if it's a bug or not,

I want to create my custom data validation, I'm trying to intercept the Records before they get updated or created.

I'm using golang to extend pocketbase.

When using the SDK my hooks work as expected, as in the UI.

But not in my custom API routes, this in an example on how I built it:

I created an API route, that extracts the data from the Body, and then saves the data as in the documentation:
https://pocketbase.io/docs/go-records/#create-new-record-without-data-validations

this is the code responsible for saving the data:

```go
type FooBar struct {
    foo string
    bar string
}

app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
	e.Router.AddRoute(echo.Route{
		Method: http.MethodPost,
		Path:   "/api/application_name/foobar/new",
		Handler: func(c echo.Context) error {
			return SaveFooBarHandler(c, app)
		},
		Middlewares: []echo.MiddlewareFunc{
			apis.RequireRecordAuth("users"),
			apis.LoadAuthContext(app),
		},
	})

	return nil
})

func SaveFooBarHandler(app *pocketbase.PocketBase, data FooBar) err {
	collection, err := app.Dao().FindCollectionByNameOrId("collection_name")
	if err != nil {
		return apis.NewApiError(500, "Something unexpected happended on our side", nil)
	}

	record := models.NewRecord(collection)
	record.Set("foo", data.foo)
	record.Set("foo", data.bar)

	if err := app.Dao().SaveRecord(record); err != nil {
		log.Default().Println("Error saving record: ", err)
		return apis.NewApiError(500, "An error ocurerred saving the data", nil)
	}
	return nil
}
``` 

I not able to find it now, but there was an option to use Dao without hooks, I'm not using that.

Maybe I have to add the middleware for the record? I'm not sure, could someone help me?

**Comment by ganigeorgiev** [comment, marked as answer]
> ... my hooks work as expected, as in the UI. But not in my custom API routes ...

What hooks? If you are listening to the `*Request` hooks then the above will not trigger it. They are fired **only** when sending API request to the specific internal endpoint to which they are attached to.

The `Dao` save operation will trigger by default only the [DB hooks](https://pocketbase.io/docs/go-event-hooks/#db-hooks) (aka. `OnModel*`). The `e.Model` in the specific db hook is an interface and if you want to access its Record value you'll need type assertion, aka. `record, ok := e.Model.(*models.Record)` (_this will be simplified with v0.23.0 as part of the [ongoing refactoring](https://github.com/pocketbase/pocketbase/discussions) changes and the introduction of proxy/projections hooks_).

_Side-note: the middleware [`LoadAuthContext`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.19/apis#LoadAuthContext) is registered by default for all routes and you can remove it from your definition._

**Reply by Ethanol48** [reply]
Thanks for the help! it worked!

**Comment by Ethanol48** [comment]
I forgot to put this code above:

```go
app.OnBeforeRecordCreateRequest("collection_name").Add(func (e) error {

// data validation
if e.Record.GetString("foo") != "foo" {
  return apis.NewBadRequest(...)
} 

return nil

})
```

This is triggered when adding a record with the SDK as I mentioned, but not by "custom" handler, is this behavior intended?

In the documentation looks like it should be called, or is when the endpoint of the database is called?

**Reply by ganigeorgiev** [reply]
Yes, as mentioned above the `*Request` hooks fire **only** when sending API request to the specific internal endpoint to which they are attached to, aka. the above will be triggered only when sending requests to the builtin `POST /api/collections/:collection/records` endpoint.

If you want to intercept any Record create operation you'll have to listen to the [DB hooks](https://pocketbase.io/docs/go-event-hooks/#db-hooks), e.g. [`OnModelBeforeCreate`](https://pocketbase.io/docs/go-event-hooks/#onmodelbeforecreate).
